core/src/main/java/com/linecorp/armeria/common/logging/LoggingDelegateBuilder.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.logging;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.Function;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.MoreObjects.ToStringHelper;
+
+import com.linecorp.armeria.client.logging.LoggingClient;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.server.logging.LoggingService;
+
+/**
+ * Builds a new {@link LoggingClient} or {@link LoggingService}.
+ */
+public class LoggingDelegateBuilder<T extends LoggingDelegateBuilder<T>> {
```

Comment:
I want to know what is the suitable name for this class.

Reply:
How about making this class as package-private and name it `AbstractLoggingBuilder`? with `ClientLoggingBuilder` and `ServiceLoggingBuilder`?  
I got this idea from `AbstractClientOptionsBuilder` :-)

Reply:
I think `LoggingDecoratorBuilder` is also a clear name since the concept of decorator is very familiar to armeria users

Reply:
Now that I think about it again, it is a bad idea since the subclasses cannot be in the same package. So, `LoggingDecoratorBuilder` is good name I think as well.
We can make the constructor as protected though. :-)